### PR TITLE
Add a development-copy build process to Atrium-based projects.

### LINKED ISSUE #27: Add symlink task for Atrium libraries
Atrium uses a core patch in order to support profile inheritance, but this inheritance does not influence library discovery. This needs to be taken into consideration when building a customized profile that extends Atrium, since these libraries within the Atrium profile directory cannot be discovered/loaded from that location.

Typically I have handled this with a custom task to symlink the libraries from the Atrium profile into `sites/all/libraries` like so:

```
// Symlink all OpenAtrium libraries into sites/all.
grunt.config(['symlink', 'oa-libraries'], {
  expand: true,
  cwd: '<%= config.buildPaths.html %>/profiles/openatrium/libraries',
  src: ['*'],
  dest: '<%= config.buildPaths.html %>/sites/all/libraries'
});
```

This could be baked into gadget, but may be require an additional prompt such as "Are you using this distro as a base for a custom profile?". There's also probably an argument for this being too specific of a use-case to include in this project, and might be more appropriate for other generator projects.

### PR DESCRIPTION
**Pull request by grayside**
In order to use pre-release/in-development Atrium code, it's more complicated than simply grabbing a development snapshot of the Atrium profile... at least if you are using your own build tooling.

This PR adds a variant build process that overrides how Drush Make is operated, assembling a git-enhanced Drupal codebase with the latest and greatest Atrium code.

Note that the current behavior in "dev-mode" places all contrib under sites/all. Need to straighten that out still.

To use this alternate reality once a project is generated for Atrium, developers will build Drupal using `grunt --dev`.

Also fixes #27

Cc @mike-potter @arithmetric

### CODE REVIEW

**Comment by mike-potter** [comment]
Forgive my ignorance of this full build process, but it looks like for the atrium-dev you do the "git clone" to get it and then you move it to profiles/openatrium.  When you run drush-make to build the dev version, where do the modules end up in the final Drupal docroot?  Do the contrib modules brought in by Atrium get properly put into /profiles/openatrium/modules?  Or do they live somewhere else?

**Comment by grayside** [comment]
As noted in the description, all contrib is ending up under sites/all. I haven't figured out to to fix that yet, are there any known issues with that structure besides registry conflicts from moving modules?

The pipeline with this change is:
1. composer install
2. code validation
3. git clone Atrium 7.x-2.x
4. Run drush via project-dev.make. This directly uses the dev makefile from build/cache/openatrium
5. Copy built codebase into build/html
6. Symlink custom code into the docroot
7. Symlink libraries from profile to sites/all
8. Copy Atrium profile into place.

The particular ordering here was lowest-hanging fruit and I have built and installed the site successfully with this sequence.

**Comment by grayside** [comment]
Botch. Nevermind.

**Comment by mike-potter** [comment]
Yeah, the problem with putting contrib into /sites is that it doesn't match how non-Phase2 Atrium sites are built, and it makes it more difficult to tell which contrib is site-specific vs belonging to a profile.  It is much better practice to put the contrib modules within the profile that actually references them in their make file rather than using the /sites folder as a dumping ground.  It also makes it easier for a specific site to override a specific contrib module for whatever reason.

This has always been the problem I have had with the symlink methods in the GDT and why I haven't been able to use those tools on Atrium.  I'd really like to figure out a good solution to this.

**Comment by mike-potter** [comment]
(and yes, this basically prevents you from using the quick-install option for Atrium as the database restore will have modules in different locations and modules such as entity api will throw fits of errors.  Moving modules that contain autoloaded PHP classes is notoriously painful)

**Comment by mike-potter** [comment]
Note that in the build.sh script that comes with Atrium, we do the drush-make like this:

  drush make --yes --no-core $DRUSH_OPTS "$ABS_CALLPATH/$MAKEFILE" --contrib-destination=profiles/openatrium

Notice the "contrib-destination" option.  That's the option we use to put the contrib stuff into the profile.  Maybe here you can use this option to put the contrib someplace and then link it to profiles/openatrium/modules/contrib? 

That might fix the contrib, but might not fix the "apps".  Where does something like oa_discussion end up?  It should be in profiles/openatrium/modules/apps, but I'm guessing on your install it ends up in sites/all/modules/apps?  Which again means you can't distinguish the apps that are part of the Atrium profile from any apps that might be added for a specific client site.

**Comment by grayside** [comment]
We could do a two-pass drush make, one for core and the project makefile, and one for this development makefile using that additional flag.

File: app/templates/drupal/openatrium/7.x/bin/grunt/atrium.js [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    make_args.push('--no-gitinfofile');
+    make_args.push('--no-cache');
+
+    // grunt-drupal-tasks does not provide a clean facilitate to override the
+    // Gruntconfig makefile.
+    make_args[1] = 'src/project-dev.make';
+    grunt.config.set('drush.make.args', make_args);
+
+    // Define the make-atrium job by duplicating the standard make configuration.
+    grunt.config.set('drush.make-atrium', grunt.config('drush.make'));
```
Comment (grayside) [inline app/templates/drupal/openatrium/7.x/bin/grunt/atrium.js:39]: Creates a duplicate make process then tailors it to target a different makefile and use a couple more flags.

File: app/templates/drupal/openatrium/7.x/bin/grunt/atrium.js [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    make_args.push('--contrib-destination=profiles/openatrium');
+    grunt.config.set('drush.make-atrium.args', make_args);
+
+    grunt.registerTask('drushmake', 'Prepare the build directory and run "drush make"', function() {
+      grunt.task.run([
+        'mkdir:init',
+        'clean:temp',
+        'drush:make',
+        // Retrieve Atrium 2.x and place in transitional codebase.
+        'shell:get-atrium-dev',
```
Comment (grayside) [inline app/templates/drupal/openatrium/7.x/bin/grunt/atrium.js:52]: Cooks atrium retrieval and the make process for it into the middle of the standard drush make process. GDT doesn't have a means to gracefully modify these steps. This alternate form is only assigned when --dev is added on grunt execution.

File: app/templates/drupal/openatrium/7.x/bin/grunt/atrium.js [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+  if (grunt.option('dev')) {
+    var make_args = grunt.config.get('drush.make.args');
+    make_args.push('--working-copy');
+    make_args.push('--no-gitinfofile');
+    make_args.push('--no-cache');
+
+    // grunt-drupal-tasks does not provide a clean facilitate to override the
+    // Gruntconfig makefile.
+    make_args[1] = 'src/project-dev.make';
+    grunt.config.set('drush.make.args', make_args);
```
Comment (grayside) [inline app/templates/drupal/openatrium/7.x/bin/grunt/atrium.js:36]: Here is where we modify the main Drush Make to do development things.

File: app/templates/drupal/openatrium/7.x/bin/grunt/atrium.js (hunk 1 above)
Comment (grayside) [inline app/templates/drupal/openatrium/7.x/bin/grunt/atrium.js:39]: This line is copying the (now modified) main drush make configuration to the atrium drush make. Next we'll modify it.

File: app/templates/drupal/openatrium/7.x/bin/grunt/atrium.js [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+    // Define the make-atrium job by duplicating the standard make configuration.
+    grunt.config.set('drush.make-atrium', grunt.config('drush.make'));
+    var make_args = grunt.config.get('drush.make-atrium.args');
+    make_args[1] = '<%= config.buildPaths.build %>/temp/profiles/openatrium/scripts/oa-drush-dev.make';
+    make_args.push('--no-core');
+    make_args.push('--contrib-destination=profiles/openatrium');
+    grunt.config.set('drush.make-atrium.args', make_args);
```
Comment (grayside) [inline app/templates/drupal/openatrium/7.x/bin/grunt/atrium.js:44]: Here we adjust the atrium make run with our changes specific to atrium.

**Comment by mike-potter** [comment]
So on line 40 you have:

var make_args = grunt.config.get('drush.make-atrium.args');

That sets the entire array.  Then on line 41 you alter the [1] arg to point to the build script.  Then you push --no-core and --contrib-dest, then you save the config.

The make_args from line 29 that has the --working-copy added gets saved on line 36 to drush.make.args.  But the "get" on line 40 is coming from drush.make-atrium.args and overwrites the entire array that is above it.

So I'm still not seeing how the --working-copy argument gets added to the drush.make-atrium.args

**Comment by grayside** [comment]
Line 39 takes all the configuration of `drush.make`, including `drush.make.args`, and initializes the `drush.make-atrium` task to have the same values. Then the specific `drush.make-atrium.args` elements are pulled out again and modified for the additional flags.

**Comment by arithmetric** [comment]
:+1:

**Pull request merged**